burstphoto/exposure/exposure.swift

```diff
@@ -28,68 +28,77 @@ func correct_exposure(_ final_texture: MTLTexture, _ white_level: Int, _ black_l
+                for j in 0..<black_levels_mean.count {
+                    black_levels_mean[j] += Double(black_level[i][j])
+                }
             }
             
-            black_level0 /= Double(exposure_bias.count)
-            black_level1 /= Double(exposure_bias.count)
-            black_level2 /= Double(exposure_bias.count)
-            black_level3 /= Double(exposure_bias.count)
-            
-        } else {
-            black_level0 = Double(black_level[exp_idx][0])
-            black_level1 = Double(black_level[exp_idx][1])
-            black_level2 = Double(black_level[exp_idx][2])
-            black_level3 = Double(black_level[exp_idx][3])
+            let count = Double(black_level.count)
+            for i in 0..<black_levels_mean.count {
+                black_levels_mean[i] /= count
+            }
         }
         
+        let black_level_min = black_levels_mean.min()!
+        let black_levels_mean_buffer = device.makeBuffer(bytes: black_levels_mean.map{Float32($0)},
+                                                         length: MemoryLayout<Float32>.size * black_levels_mean.count)!
+        
         let command_buffer = command_queue.makeCommandBuffer()!
         command_buffer.label = "Correct Exposure"
         let command_encoder = command_buffer.makeComputeCommandEncoder()!
-        let state = (exposure_control=="Curve0EV" || exposure_control=="Curve1EV") ? correct_exposure_state : correct_exposure_linear_state
-        command_encoder.setComputePipelineState(state)
-        let threads_per_grid = MTLSize(width: final_texture.width, height: final_texture.height, depth: 1)
-        let threads_per_thread_group = get_threads_per_thread_group(state, threads_per_grid)
+        let state: MTLComputePipelineState
        
         if (exposure_control=="Curve0EV" || exposure_control=="Curve1EV") {
-            let color_factor_mean = 0.25*(color_factors[ref_idx][0]+2.0*color_factors[ref_idx][1]+color_factors[ref_idx][2])
+            state = correct_exposure_state
+            
+            let black_level_mean = Double(black_levels_mean.reduce(0, +)) / Double(black_levels_mean.count)
+            let color_factor_mean: Double
+            if (mosaic_pattern_width == 6) {
+                color_factor_mean = (8.0*color_factors[ref_idx][0] + 20.0*color_factors[ref_idx][1] + 8.0*color_factors[ref_idx][2]) / 36.0
+            } else if (mosaic_pattern_width == 2) {
+                color_factor_mean = (    color_factors[ref_idx][0] +  2.0*color_factors[ref_idx][1] +     color_factors[ref_idx][2]) /  4.0
+            } else {
+                color_factor_mean = (    color_factors[ref_idx][0] +      color_factors[ref_idx][1] +     color_factors[ref_idx][2]) /  3.0
+            }
             
             // the blurred texture serves as an approximation of local luminance
             final_texture_blurred = blur(final_texture, with_pattern_width: 1, using_kernel_size: 1)
```

Comment:
For XTrans, the kernel size should be 2 instead of 1. This ensures that each pixel in the blurred texture has a comparable contribution of R, G, B pixels. For Bayer, the kernel size is correct.

Reply:
Good catch